### PR TITLE
✨Added databricks source to viadot

### PR DESCRIPTION
**Pull request by afraijat**
## Summary

Added databricks source to viadot library

## Importance

Added databricks source to viadot library with the following functionalities:
1) Creating, reading and deleting tables
2) Converting to and from Spark DataFrames
3) Create and delete schemas
4) Upserting records
5) Full-refresh
6) Printing data schema of a table

## Checklist

This PR:

- [X] adds new tests (if appropriate)
- [X] updates docstrings for any new functions or function arguments (if appropriate)
- [X] updates `CHANGELOG.md` with a summary of the changes

### CODE REVIEW

File: viadot/sources/databricks.py [hunk 1]
```diff
@@ -0,0 +1,350 @@
+import json
+from pandas import DataFrame
```
Comment (Trymzet) [inline viadot/sources/databricks.py:2]: This import is confusing as you're dealing with two types of DataFrames in this file. This should be imported as eg. `pd.DataFrame` to always know which DataFrame class we're dealing with.

File: viadot/sources/databricks.py [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+    def connect(self):
+        if self.env == "QA":
+            self.spark = SparkSession.builder.getOrCreate()
+        else:
+            self.spark = self._get_spark_session(self.env)
+    
+    def to_df(self, list):
+        return pd.DataFrame(list)
+    
+    def to_json(sparkDf: pyspark.sql.dataframe.DataFrame):
```
Comment (Trymzet) [inline viadot/sources/databricks.py:73]: I would import this as `spark.DataFrame` for clarity. Then you have `pd.DataFrame` and `spark.DataFrame`.

File: viadot/sources/databricks.py [hunk 3]
```diff
@@ -0,0 +1,350 @@
+import json
+from pandas import DataFrame
+from viadot.config import local_config
+from viadot.sources.base import Source
+from delta.tables import *
+import pandas as pd
+import pyspark
+from pyspark import SparkConf, SparkContext
+import time
+# from viadot.sources import Base
```
Comment (Trymzet) [inline viadot/sources/databricks.py:10]: Remove

File: viadot/sources/databricks.py [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+class Databricks(Source):
```
Comment (Trymzet) [inline viadot/sources/databricks.py:11]: Use Black to format the code; there should be two spaces before a class definition.

File: viadot/sources/databricks.py [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+        ```
+        Returns:
+            pyspark.sql.dataframe.DataFrame: The converted Spark DataFrame
+        """
+        sparkDF = self.spark.createDataFrame(df)
+        return sparkDF
+    
+    def spark_to_pandas(self, sparkDf: pyspark.sql.dataframe.DataFrame):
+        """
+        Convert Spark DataFrame to Pandas DataFrame
```
Comment (Trymzet) [inline viadot/sources/databricks.py:98]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+        Example:
+        ```python
+        from viadot.sources import Databricks
+        databricks = Databricks()
+        list = [{"some":"data"}]
+        sparkDf = databricks.to_spark_df(df)
+        pandasDf = spark_to_pandas(sparkDf)
+        ```
+        Returns:
+            pyspark.sql.dataframe.DataFrame: The converted Spark DataFrame
```
Comment (Trymzet) [inline viadot/sources/databricks.py:112]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+        """
+        return sparkDf.toPandas()
+    
+    def query(self, query: str):
+        """
+        Execute an SQL query
```
Comment (Trymzet) [inline viadot/sources/databricks.py:118]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+            query (str): An SQL query
+
+        Example:
+        ```python
+        from viadot.sources import Databricks
+        databricks = Databricks()
+        queryResult = databricks.query("SELECT * FROM schema.table_1")
+        ```
+        Returns:
+            DataFrame: A Pandas DataFrame containing the query's results (if available)
```
Comment (Trymzet) [inline viadot/sources/databricks.py:130]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+        """
+        result = self.spark.sql(query)
+        return self.spark_to_pandas(result)
+    
+    def read_table(self, table_name: str):
+        """
+        Read the contents of a given table
```
Comment (Trymzet) [inline viadot/sources/databricks.py:137]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+
+        Args:
+            table_name (str): Name of an existing table
```
Comment (Trymzet) [inline viadot/sources/databricks.py:140]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+
+        Example:
+        ```python
+        from viadot.sources import Databricks
+        databricks = Databricks()
+        tableData = databricks.read_table("schema.table_1")
+        ```
+        Returns:
+            DataFrame: A Pandas DataFrame containing the requested table's data
```
Comment (Trymzet) [inline viadot/sources/databricks.py:149]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+        """
+        result = self.spark.sql("select * from "+ table_name)
+        return self.spark_to_pandas(result)
+
+    def create_table(self, table_name: str, df: DataFrame):
+        """
+        Write a new table using a given Pandas DataFrame
```
Comment (Trymzet) [inline viadot/sources/databricks.py:156]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+
+        Args:
+            table_name (str): Name of the new table to be created
+            df (DataFrame): DataFrame to be written as a table
```
Comment (Trymzet) [inline viadot/sources/databricks.py:160]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+
+        Example:
+        ```python
+        from viadot.sources import Databricks
+        databricks = Databricks()
+        df = [{"id":"1", "name":"Joe"}]
+        new_table = databricks.create_table("schema.table_1", df)
+        ```
+        Returns:
+            DataFrame: A Pandas DataFrame containing the newly created table's data
```
Comment (Trymzet) [inline viadot/sources/databricks.py:170]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 15]
```diff
@@ -0,0 +1,350 @@
...
+        data.createOrReplaceTempView("new_table")
+
+        self.query("CREATE TABLE " + table_name +" USING DELTA AS SELECT * FROM new_table;")
+
+        print("Table " +table_name+ " created.")
+        return self.read_table(table_name)    
+
+    def delete_table(self, table_name: str):
+        """
+        Delete an existing table
```
Comment (Trymzet) [inline viadot/sources/databricks.py:182]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+
+        Args:
+            table_name (str): Name of the table to be overwritten
+            df (DataFrame): DataFrame to be used to overwrite the table
```
Comment (Trymzet) [inline viadot/sources/databricks.py:186]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+        databricks = Databricks()]
+        databricks.delete_table("schema.table_1")
+        ```
+        """
+        self.spark.sql("DROP TABLE "+ table_name)
+        print("Table deleted.")
+
+    def full_refresh(self, table_name: str, df: DataFrame):
+        """
+        Overwrite an existing table with data from a Pandas DataFrame
```
Comment (Trymzet) [inline viadot/sources/databricks.py:200]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 18]
```diff
@@ -0,0 +1,350 @@
...
+
+        Args:
+            table_name (str): Name of the table to be overwritten
+            df (DataFrame): DataFrame to be used to overwrite the table
```
Comment (Trymzet) [inline viadot/sources/databricks.py:204]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 19]
```diff
@@ -0,0 +1,350 @@
...
+
+        Example:
+        ```python
+        from viadot.sources import Databricks
+        databricks = Databricks()
+        df = [{"id":"1", "name":"Joe"}]
+        new_table = databricks.full_refresh("schema.table_1", df)
+        ```
+        Returns:
+            DataFrame: A Pandas DataFrame containing the new data in the overwitten table
```
Comment (Trymzet) [inline viadot/sources/databricks.py:214]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 20]
```diff
@@ -0,0 +1,350 @@
...
+        data = self.to_spark_df(df)
+        data.write.format("delta").mode("overwrite").saveAsTable(table_name)
+        return self.read_table(table_name)
+        
+    # TODO: add a method for extracting a usable JSON schema for the sets
+    # TODO: Parse the matchedSet as a notMatchedSet
+    def upsert(self, table_name: str, df: DataFrame, matchedSet: dict, notMatchedSet: dict, pk: str):
+        """
+        Upsert the data to an existing table. If the record exists, the function updates the record in the table.
+        If it doesn't exist, the function creates a new record in the table
```
Comment (Trymzet) [inline viadot/sources/databricks.py:225]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 21]
```diff
@@ -0,0 +1,350 @@
...
+
+        Args:
+            table_name (str): Name of the table to perform the upsert operation on
+            df (DataFrame): DataFrame to be used to upsert the table
+            matchedSet (dict): A dictionary defining the changes that should take place in case a matching record is found
+            notMatchedSet(dict): A dictionary defining the fields that will be added in case no records exist
+            pk (str): The primary key that should be joined on
```
Comment (Trymzet) [inline viadot/sources/databricks.py:232]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 22]
```diff
@@ -0,0 +1,350 @@
...
+    def upsert(self, table_name: str, df: DataFrame, matchedSet: dict, notMatchedSet: dict, pk: str):
+        """
+        Upsert the data to an existing table. If the record exists, the function updates the record in the table.
+        If it doesn't exist, the function creates a new record in the table
+
+        Args:
+            table_name (str): Name of the table to perform the upsert operation on
+            df (DataFrame): DataFrame to be used to upsert the table
+            matchedSet (dict): A dictionary defining the changes that should take place in case a matching record is found
+            notMatchedSet(dict): A dictionary defining the fields that will be added in case no records exist
```
Comment (Trymzet) [inline viadot/sources/databricks.py:231]: Remove camelcase

File: viadot/sources/databricks.py [hunk 23]
```diff
@@ -0,0 +1,350 @@
...
+        data = [{"Id": "KVSzUaILfQZXDb", "wrong": "Updated3!EHNYKjSZsiy", "Name": "Upsert-Black", "FirstName": "Updated", "LastName": "Carter2", "ContactEMail": "Adam.Carter@TurnerBlack.com", "MailingCity": "Updated!Jamesport"}]
+        match = {"Id": "main.Id",
+                "AccountId": "updates.AccountId",
+                "Name": "updates.Name",
+                "FirstName": "updates.FirstName",
+                "LastName": "updates.LastName",
+                "ContactEMail": "updates.ContactEMail",
+                "MailingCity": "main.MailingCity"}
+
+        notMatch = {
```
Comment (Trymzet) [inline viadot/sources/databricks.py:246]: remove camelcase

File: viadot/sources/databricks.py [hunk 24]
```diff
@@ -0,0 +1,350 @@
...
+            " WHEN MATCHED THEN UPDATE SET " + matchedString + " WHEN NOT MATCHED THEN INSERT (" + notMatchedKeys
+            + ") VALUES (" + notMatchedValues + ")")
+
+            print("Data upserted successfully.")
+        else:
+            print("Columns mismatch. Please adhere to the column names and data types used in the source table.")
+    
+    def create_schema(self, schemaName: str):
+        """
+        Create a schema for storing tables
```
Comment (Trymzet) [inline viadot/sources/databricks.py:302]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 25]
```diff
@@ -0,0 +1,350 @@
...
+
+        Args:
+            schemaName (str): Name of the new schema to be created
```
Comment (Trymzet) [inline viadot/sources/databricks.py:305]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 26]
```diff
@@ -0,0 +1,350 @@
...
+        databricks = Databricks()
+        databricks.create_schema("schema_1")
+        ```
+        """
+        self.spark.sql("CREATE SCHEMA IF NOT EXISTS " + schemaName)
+        print("Schema "+ schemaName + " created.")
+
+    def delete_schema(self, schemaName: str):
+        """
+        Delete a schema
```
Comment (Trymzet) [inline viadot/sources/databricks.py:319]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 27]
```diff
@@ -0,0 +1,350 @@
...
+
+        Args:
+            schemaName (str): Name of the schema to be deleted
```
Comment (Trymzet) [inline viadot/sources/databricks.py:322]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 28]
```diff
@@ -0,0 +1,350 @@
...
+        databricks = Databricks()
+        databricks.delete_schema("schema_1")
+        ```
+        """
+        self.spark.sql("DROP SCHEMA " + schemaName)
+        print("Schema "+ schemaName + " deleted.")
+
+    def print_schema(self, table_name: str):
+        """
+        Print the details of a table's data schema
```
Comment (Trymzet) [inline viadot/sources/databricks.py:336]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 29]
```diff
@@ -0,0 +1,350 @@
...
+
+        Args:
+            table_name (str): Name of the table whose schema will be printed
```
Comment (Trymzet) [inline viadot/sources/databricks.py:339]: Use dots at the end of sentences

File: tests/integration/tasks/test_databricks.py [hunk 30]
```diff
@@ -0,0 +1,98 @@
...
+    table_name = "raw.test_table"
+    match = {"Id": "main.Id",
+            "AccountId": "updates.AccountId",
+            "Name": "updates.Name",
+            "FirstName": "updates.FirstName",
+            "LastName": "updates.LastName",
+            "ContactEMail": "updates.ContactEMail",
+            "MailingCity": "main.MailingCity"}
+
+    notMatch = {
```
Comment (Trymzet) [inline tests/integration/tasks/test_databricks.py:82]: Remove camelcase

File: viadot/sources/databricks.py [hunk 31]
```diff
@@ -0,0 +1,350 @@
...
+    
+    def to_df(self, list):
+        return pd.DataFrame(list)
+    
+    def to_json(sparkDf: pyspark.sql.dataframe.DataFrame):
+        return sparkDf.toJSON()
+    
+    def to_spark_df(self, df: DataFrame):
+        """
+        Convert Pandas DataFrame to Spark DataFrame
```
Comment (Trymzet) [inline viadot/sources/databricks.py:78]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 32]
```diff
@@ -0,0 +1,350 @@
...
+
+        Args:
+            df (DataFrame): Pandas DataFrame to be converted to Spark DataFrame.
+
+        Example:
+        ```python
+        from viadot.sources import Databricks
+        databricks = Databricks()
+        list = [{"some":"data"}]
+        sparkDf = databricks.to_spark_df(list)
```
Comment (Trymzet) [inline viadot/sources/databricks.py:88]: The example will have to be adjusted once the method is adjusted as per comments.

File: viadot/sources/databricks.py [hunk 33]
```diff
@@ -0,0 +1,350 @@
...
+        ```
+        Returns:
+            pyspark.sql.dataframe.DataFrame: The converted Spark DataFrame
+        """
+        sparkDF = self.spark.createDataFrame(df)
+        return sparkDF
```
Comment (Trymzet) [inline viadot/sources/databricks.py:94]: Remove camel case

File: viadot/sources/databricks.py [hunk 34]
```diff
@@ -0,0 +1,350 @@
...
+            pyspark.sql.dataframe.DataFrame: The converted Spark DataFrame
+        """
+        return sparkDf.toPandas()
+    
+    def query(self, query: str):
+        """
+        Execute an SQL query
+
+        Args:
+            query (str): An SQL query
```
Comment (Trymzet) [inline viadot/sources/databricks.py:121]: Use dots at the end of sentences

File: viadot/sources/databricks.py [hunk 35]
```diff
@@ -0,0 +1,350 @@
...
+        databricks = Databricks()
+        queryResult = databricks.query("SELECT * FROM schema.table_1")
+        ```
+        Returns:
+            DataFrame: A Pandas DataFrame containing the query's results (if available)
+        """
+        result = self.spark.sql(query)
+        return self.spark_to_pandas(result)
+    
+    def read_table(self, table_name: str):
```
Comment (Trymzet) [inline viadot/sources/databricks.py:135]: Not sure if we need that if we have `query()` and `to_df()` already?

**Review by Trymzet — CHANGES_REQUESTED** [review]
I'm probably not finished but I need the code to be a bit clearer to review further; there's plenty to work on anyway.

File: viadot/sources/databricks.py [hunk 36]
```diff
@@ -44,146 +57,160 @@ def _get_spark_session(self, env: str = None):
...
+            str: A JSON string containing the requested table's data.
         """
-        return sparkDf.toPandas()
-    
-    def query(self, query: str):
+        result = self.sc.sql("select * from " + table_name)
+        df = self._spark_df_to_pandas_df(result)
+        if df.empty:
+            self._handle_if_empty(if_empty)
+        return df.to_json()
```
Comment (Trymzet) [inline viadot/sources/databricks.py:108]: The purpose of `to_json()` is to offer a more flexible way to consume data (not necessarily a pandas DataFrame). If you need to first load it to a DataFrame anyway then this function doesn't make sense -- user can just use `Databricks.to_df()`.

File: viadot/sources/databricks.py [hunk 37]
```diff
@@ -44,146 +57,160 @@ def _get_spark_session(self, env: str = None):
...
 
         Example:
         ```python
         from viadot.sources import Databricks
         databricks = Databricks()
-        tableData = databricks.read_table("schema.table_1")
+        list = [{"id":"1", "name":"Joe"}]
+        df = pd.DataFrame(list)
+        spark_df = databricks._pandas_df_to_spark_df(df)
+        pandas_df = _spark_df_to_pandas_df(spark_df)
```
Comment (Trymzet) [inline viadot/sources/databricks.py:146]: Should be: `databricks._spark_df_to_pandas_df(spark_df)`

File: viadot/sources/databricks.py [hunk 38]
```diff
@@ -44,146 +57,160 @@ def _get_spark_session(self, env: str = None):
...
+            spark.DataFrame: The converted Spark DataFrame.
         """
-        result = self.spark.sql(query)
-        return self.spark_to_pandas(result)
-    
-    def read_table(self, table_name: str):
+        spark_df = self.sc.createDataFrame(df)
+        return spark_df
+
+    def _spark_df_to_pandas_df(self, spark_df: spark.DataFrame):
```
Comment (Trymzet) [inline viadot/sources/databricks.py:132]: Missing return type annotation.

**Review by Trymzet — CHANGES_REQUESTED** [review]

File: tests/integration/tasks/test_databricks.py [hunk 39]
```diff
@@ -305,35 +305,40 @@ def test_upsert():
...
-    match = {
-        "Id": "main.Id",
-        "AccountId": "updates.AccountId",
-        "Name": "updates.Name",
-        "FirstName": "updates.FirstName",
-        "LastName": "updates.LastName",
-        "ContactEMail": "updates.ContactEMail",
-        "MailingCity": "main.MailingCity",
-    }
+    match = "Id= main.Id, AccountId= updates.AccountId, Name= updates.Name, FirstName= updates.FirstName, LastName= updates.LastName, ContactEMail= updates.ContactEMail, MailingCity= main.MailingCity"
```
[the remark this answers is not in the document]
Comment (afraijat) [reply tests/integration/tasks/test_databricks.py:333]: 1. Noted and fixed

2. Previous implementations used dictionaries, and the final result needs to be a string. Both got rejected (first because the dictionaries were converted to a string, second because it was different from `utils.build_merge_query()`) and as such, I opted to using a string directly.

3. `utils.build_merge_query()` assumes that the user wants to update all the fields in every single upsert operation; it does not give them the option of selectively updating certain fields while allowing other fields to retain their old values. This function allows them to specify which fields will be updated, and which ones will be left intact. For example:
` match="Id=main.Id, AccountId=updates.AccountId, Name=updates.Name, FirstName=updates.FirstName, LastName=updates.LastName, ContactEMail=updates.ContactEMail, MailingCity=main.MailingCity"`
Here, we specify that we want to insert the value of `MailingCity` corresponding to the `main` table, which is the existing/main table that we read from Databricks; in other words, this field won't be updated. The other columns (besides the primary key, but that's irrelevant to the end result) have their values updated using the values from the `updates` temporary view, which is the staging table.

4. For a. the reason described in point 3 and b. The query used to retrieve the column names in `utils.build_merge_query()` does not work with Databricks/SparkSQL; it requires a different SQL statement

5. Noted and fixed

File: tests/integration/tasks/test_databricks.py [hunk 40]
```diff
@@ -305,35 +305,40 @@ def test_upsert():
...
+    data = pd.DataFrame(data)
+    query = databricks.build_merge_query(table_name, "Id", data, match)
 
     df = pd.DataFrame(source_data)
     df = df.sort_values(by="Id")
     df = df.reset_index(drop=True)
 
-    databricks.upsert(table_name, upsert_data, match, "Id")
+    databricks.query(query)
```
[the remark this answers is not in the document]
Comment (afraijat) [reply tests/integration/tasks/test_databricks.py:341]: 1. `databricks.upsert()` was using `databricks.query()` before to run the upsert statement internally; the the difference is now the query building and execution are split up.

2. `databricks.build_merge_query()` builds the upsert query only; it does not run it. We need to run the query/statement somehow. If the name `query` is confusing, I can rename the `query` function to `sql`; this way it would be clearer that it's for general SQL statements, both querying data and performing operations (ex. create table/schema, delete table etc.)

File: viadot/sources/databricks.py [hunk 41]
```diff
@@ -245,82 +245,48 @@ def full_refresh(self, table_name: str, df: pd.DataFrame):
...
-                + " VALUES "
-                + columns
-            )
-
-            print("Data upserted successfully.")
-        else:
-            print(
-                "Columns mismatch. Please adhere to the column names and data types used in the source table."
-            )
+        upsertData.createOrReplaceTempView("updates")
```
[the remark this answers is not in the document]
Comment (afraijat) [reply viadot/sources/databricks.py:272]: As mentioned in the other comment, the function in `utils.py` uses a different SQL statement for retrieving the column names and does not support selectively updating fields of a record; it cannot be reused without changing its implementation

**Review by Trymzet — CHANGES_REQUESTED** [review]

File: viadot/sources/databricks.py [hunk 42]
```diff
@@ -0,0 +1,396 @@
...
+
+    def run(
+        self, query: str, fetch_type: Literal["record", "dataframe"] = "record"
+    ) -> Union[spark.DataFrame, pd.DataFrame, bool]:
+        """
+        Execute an SQL query.
+
+        Args:
+            query (str): The query to execute.
+            fetch_type (Literal[, optional): How to return the data: either
```
Comment (Trymzet) [inline viadot/sources/databricks.py:146]: Typo in the typing

File: viadot/sources/databricks.py [hunk 43]
```diff
@@ -0,0 +1,396 @@
...
+            schema_name (str): Name of the new schema to be created.
+
+        Example:
+        ```python
+        from viadot.sources import Databricks
+        databricks = Databricks()
+        databricks.create_schema("schema_1")
+        ```
+        """
+        self.session.sql(f"CREATE SCHEMA IF NOT EXISTS {schema_name}")
```
Comment (Trymzet) [inline viadot/sources/databricks.py:348]: Should use `self.run()`

File: viadot/sources/databricks.py [hunk 44]
```diff
@@ -0,0 +1,396 @@
...
+            schema_name (str): Name of the schema to be deleted.
+
+        Example:
+        ```python
+        from viadot.sources import Databricks
+        databricks = Databricks()
+        databricks.delete_schema("schema_1")
+        ```
+        """
+        self.session.sql(f"DROP SCHEMA {schema_name}")
```
Comment (Trymzet) [inline viadot/sources/databricks.py:365]: Should use `self.run()`

File: viadot/sources/databricks.py [hunk 45]
```diff
@@ -0,0 +1,396 @@
...
+        databricks = Databricks()
+        databricks.discover_schema("table_1")
+        ```
+        Returns:
+            schema: A dictionary containing the schema details of the table.
+        """
+        result = self.run(f"DESCRIBE {table_name}", "dataframe")
+
+        # Delete the last 3 records, which are boilerplate for Delta Tables
+        names = result["col_name"].values.tolist()
```
Comment (Trymzet) [inline viadot/sources/databricks.py:388]: I think `col_names` or `column_names` would be a more precise name.

File: viadot/sources/databricks.py [hunk 46]
```diff
@@ -0,0 +1,396 @@
...
+        del names[-3:]
+
+        data_types = result["data_type"].values.tolist()
+        del data_types[-3:]
```
Comment (Trymzet) [inline viadot/sources/databricks.py:392]: In Python it's preferred to slice rather than modify arrays. You can do `names = names[:-2]`.

File: viadot/sources/databricks.py [hunk 47]
```diff
@@ -0,0 +1,396 @@
...
+
+        schema = {names[i]: data_types[i] for i in range(len(names))}
```
Comment (Trymzet) [inline viadot/sources/databricks.py:394]: It seems like you could use `zip()` here (I think `dict(zip(col_names, dtypes))` should work).

File: viadot/sources/databricks.py [hunk 48]
```diff
@@ -0,0 +1,396 @@
...
+        Example:
+        ```python
+        from viadot.sources import Databricks
+        databricks = Databricks()
+        databricks.discover_schema("table_1")
+        ```
+        Returns:
+            schema: A dictionary containing the schema details of the table.
+        """
+        result = self.run(f"DESCRIBE {table_name}", "dataframe")
```
Comment (Trymzet) [inline viadot/sources/databricks.py:385]: It seems like you make your life harder by fetching this is a DataFrame, can't we fetch this as a list or dict?

File: viadot/sources/databricks.py [hunk 49]
```diff
@@ -0,0 +1,396 @@
...
+            result = True
+
+        return result
+
+    def _check_if_table_exists(self, table_name: str) -> bool:
+        split_name = table_name.split(".")
+        does_exist = self.session._jsparkSession.catalog().tableExists(
+            split_name[0], split_name[1]
+        )
+        return does_exist
```
Comment (Trymzet) [inline viadot/sources/databricks.py:186]: Refer back to the comment about splitting table and schema as separate parameters.

**Review by Trymzet — CHANGES_REQUESTED** [review]

File: viadot/sources/databricks.py (hunk 48 above)
Comment (afraijat) [reply viadot/sources/databricks.py:385]: Unfortunately no, since it's a Spark sql statement running, it's not possible to directly fetch it as a list or dict.

Comment (Trymzet) [reply viadot/sources/databricks.py:385]: Then why do we have the `fetch_type` parameter at all? 😅 If you don't use it then you can remove it. Then this stays the same just without having to pass "dataframe" as the argument to `fetch_type`.

Comment (afraijat) [reply viadot/sources/databricks.py:385]: I'm not sure what you mean. By default, it fetches it as a Spark DataFrame (the default record type), but I want it to be returned as a Pandas DataFrame for the ease of manipulating it.

**Comment by Trymzet** [comment]
I mean if we never need anything else than a pandas dataframe then you can
remove the fetch type parameter.

**Comment by Trymzet** [comment]
If we do use the Spark dataframe as well then you're right we need to
specifiy the return type.

However then pls change the name as you have "dataframe" which is the most
confusing name possible since you literally can only return two dataframe
types (spark and pandas).

**Comment by Trymzet** [comment]
Pls also add .prefect to .gitignore as discussed

**Comment by afraijat** [comment]
> 

I checked, `.prefect` is already in `.gitignore`. I'm not sure how the folder got pushed.

File: tests/integration/tasks/test_databricks.py [hunk 50]
```diff
@@ -0,0 +1,183 @@
...
+            "FirstName": "Melody",
+            "LastName": "Cook",
+            "ContactEMail": "Melody.Cook@ScottMerritt.com",
+            "MailingCity": "Elizabethfurt",
+        }
+    ]
+
+    full_refresh_df = pd.DataFrame(full_refresh_data)
+
+    did_func_work = databricks.insert_into(
```
Comment (lzuchowska) [inline tests/integration/tasks/test_databricks.py:116]: Same case as the previous comment.

File: tests/integration/tasks/test_databricks.py [hunk 51]
```diff
@@ -0,0 +1,183 @@
...
+
+    table = "test_upsert"
+    fqn = f"{schema}.{table}"
+    primary_key = "Id"
+
+    databricks.create_table_from_pandas(schema, table, df)
+
+    upsert_df = pd.DataFrame(upsert_data)
+
+    did_func_work = databricks.insert_into(
```
Comment (lzuchowska) [inline tests/integration/tasks/test_databricks.py:151]: Same as the previous comment

File: viadot/sources/databricks.py [hunk 52]
```diff
@@ -0,0 +1,436 @@
...
+    def __init__(
+        self, env: str = "DEV", credentials: Dict[str, Any] = None, *args, **kwargs
+    ):
+        self.env = env
+
+        self.credentials = credentials or local_config.get("DATABRICKS", {}).get(env)
+
+        self.connect()
+
+    def _get_spark_session(self, env: str = "DEV"):
```
Comment (lzuchowska) [inline viadot/sources/databricks.py:52]: This method created a new spark session, reads config, adjusts it and returns a `new_spark`. Shouldn't it be named more like `_create_spark_session` than `_get_spark_session`?

Moreover, we could name `new_config` just as `config`, as there is no *old* config or any other local config variable.

File: viadot/sources/databricks.py [hunk 53]
```diff
@@ -0,0 +1,436 @@
...
+                f"Only the values {allowed_fetch_type_values} are allowed for 'fetch_type'"
+            )
+
+        query_clean = query.upper().strip()
+        query_keywords = ["SELECT", "SHOW", "PRAGMA", "DESCRIBE"]
+
+        query_result = self.session.sql(query)
+
+        if any(query_clean.startswith(word) for word in query_keywords):
+            if fetch_type == "record":
```
Comment (lzuchowska) [inline viadot/sources/databricks.py:194]: Correct me if I'm wrong but possible values for `fetch_type` are `spark` and `pandas` as far as I can see, so this statement would never execute.

File: viadot/sources/databricks.py [hunk 54]
```diff
@@ -0,0 +1,436 @@
...
+        Delete an existing table.
+
+        Args:
+            schema (str): Name of the schema
+            table (str): Name of the new table to be created.
+
+        Example:
+        ```python
+        from viadot.sources import Databricks
+        databricks = Databricks()]
```
Comment (lzuchowska) [inline viadot/sources/databricks.py:255]: Please fix a typo at `Databricks()]`

File: viadot/sources/databricks.py [hunk 55]
```diff
@@ -0,0 +1,436 @@
...
+    ):
+        """
+        Insert data into a table.
+
+        Args:
+            schema (str): Name of the schema
+            table (str): Name of the new table to be created.
+            df (pd.DataFrame): DataFrame with the data to be inserted into the table.
+            primary_key (str, Optional): Needed only when updating the data, the primary key on which the data will be joined.
+            if_exists (str, Optional): Which operation to run with the data. Allowed operations are: replace, append and update.
```
Comment (lzuchowska) [inline viadot/sources/databricks.py:344]: `if_exists` can also have a value `fail`, please correct that

**Review by lzuchowska — CHANGES_REQUESTED** [review]
To summarize:

- [ ] `did_func_work` naming change in line 80. in `test_databricks.py`
- [ ]  `did_func_work` naming change in line 116. in `test_databricks.py`
- [ ]  `did_func_work` naming change in line 151. in `test_databricks.py`
- [ ] please add some tests for negative cases - when insert *shouldn't* happen, make sure those cases are covered up too.
- [ ]  name change in `databricks.py` in line 69.
- [ ] rename method name and variables inside `_get_spark_session()`
- [ ] `if fetch_type == "record"` statement correction in `run` method
- [ ] fix typo in docstrings in `drop_table()` method
- [ ] change the `fetch_type` check in `run()` method
- [ ] fix docstrings in `insert_into()` method
- [ ] change ifs in `insert_into()` method
- [ ] Remove comment above `discover_schema()` method

**Comment by lzuchowska** [comment]
Looks fine to me!

**Review by lzuchowska — APPROVED** [review]
